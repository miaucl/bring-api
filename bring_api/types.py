"""Bring API types."""
from enum import Enum
from typing import List, TypedDict


class BringList(TypedDict):
    """A list class. Represents a single list."""

    listUuid: str
    name: str
    theme: str


class BringPurchase(TypedDict):
    """A purchase class. Represents a single item."""

    uuid: str
    itemId: str
    specification: str


class BringListItemDetails(TypedDict):
    """An item details class.

    Includes several details of an item in the context of a list.
    Caution: This does not have to be an item that is currently marked as 'to buy'.
    """

    uuid: str
    itemId: str
    listUuid: str
    userIconItemId: str
    userSectionId: str
    assignedTo: str
    imageUrl: str


class BringAuthResponse(TypedDict):
    """An auth response class."""

    uuid: str
    publicUuid: str
    email: str
    name: str
    photoPath: str
    bringListUUID: str
    access_token: str
    refresh_token: str
    token_type: str
    expires_in: int


class BringListResponse(TypedDict):
    """A list response class."""

    lists: List[BringList]


class BringItemsResponse(TypedDict):
    """An items response class."""

    uuid: str
    status: str
    purchase: List[BringPurchase]
    recently: List[BringPurchase]


class BringListItemsDetailsResponse(List[BringListItemDetails]):
    """A response class of a list of item details."""


class BringNotificationType(Enum):
    """Notification type.

    GOING_SHOPPING: "I'm going shopping! - Last chance for adjustments"
    CHANGED_LIST: "List changed - Check it out"
    SHOPPING_DONE: "Shopping done - you can relax"
    URGENT_MESSAGE: "Breaking news - Please get {itemName}!
    """

    GOING_SHOPPING = "GOING_SHOPPING"
    CHANGED_LIST = "CHANGED_LIST"
    SHOPPING_DONE = "SHOPPING_DONE"
    URGENT_MESSAGE = "URGENT_MESSAGE"


class BringNotificationsConfigType(TypedDict):
    """A notification config."""

    arguments: List[str]
    listNotificationType: str
    senderPublicUserUuid: str


class BringUserSettingsEntry(TypedDict):
    """A user settings class. Represents a single user setting."""

    key: str
    value: str


class BringUserListSettingEntry(TypedDict):
    """A user list settings class. Represents a single list setting."""

    listUuid: str
    usersettings: List[BringUserSettingsEntry]


class BringUserSettingsResponse(TypedDict):
    """A user settings response class."""

    usersettings: List[BringUserSettingsEntry]
    userlistsettings: List[BringUserListSettingEntry]


class BringSyncCurrentUserResponse(TypedDict):
    """A sync current user response class."""

    email: str
    name: str
    publicUuid: str
    userLocale: dict[str, str]
    userUuid: str
