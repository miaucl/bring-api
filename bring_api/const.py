"""Constants for bring-api."""
from typing import Final

API_BASE_URL: Final = "https://api.getbring.com/rest/"
DEFAULT_HEADERS: Final = {
    "Authorization": "Bearer",
    "X-BRING-API-KEY": "cof4Nc6D8saplXjE3h3HXqHH8m7VU2i1Gs0g85Sp",
    "X-BRING-CLIENT": "android",
    "X-BRING-APPLICATION": "bring",
    "X-BRING-COUNTRY": "DE",
    "X-BRING-USER-UUID": "",
}

LOCALES_BASE_URL: Final = "https://web.getbring.com/locale/"
BRING_SUPPORTED_LOCALES: Final = [
    "en-AU",
    "de-DE",
    "fr-FR",
    "it-IT",
    "en-CA",
    "nl-NL",
    "nb-NO",
    "pl-PL",
    "pt-BR",
    "ru-RU",
    "sv-SE",
    "de-CH",
    "fr-CH",
    "it-CH",
    "es-ES",
    "tr-TR",
    "en-GB",
    "en-US",
    "hu-HU",
    "de-AT",
]

BRING_DEFAULT_LOCALE: Final = "de-CH"
