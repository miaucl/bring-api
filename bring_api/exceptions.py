"""Bring API exceptions."""


class BringException(Exception):
    """General exception occurred."""


class BringAuthException(BringException):
    """When an authentication error is encountered."""


class BringRequestException(BringException):
    """When the HTTP request fails."""


class BringParseException(BringException):
    """When parsing the response of a request fails."""


class BringEMailInvalidException(BringException):
    """When checkemail returns emailValid false ."""


class BringUserUnknownException(BringException):
    """When checkemail returns userExists false."""


class BringTranslationException(BringException):
    """When translating an article fails."""
