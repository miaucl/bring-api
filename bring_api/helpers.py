"""Helpers for the Bring API."""

import json
from typing import cast


def headers_serialize(headers: dict[str, str]) -> str:
    """Serialize all headers from the login session to a string for persistency.

    Parameters
    ----------
    headers
        The headers in the session to persist

    str
        The serialized data to persist

    """
    return json.dumps(headers)


def headers_deserialize(headers: str) -> dict[str, str]:
    """Deserialize all headers from the login session from a persisted string.

    Parameters
    ----------
    headers
        The persisted session headers as string

    Returns
    -------
    dict[str, str]
        The session cookies

    """
    return cast(dict[str, str], json.loads(headers))
