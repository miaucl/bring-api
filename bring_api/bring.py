"""Bring api implementation."""
import asyncio
from json import JSONDecodeError
import logging
import traceback
from typing import Optional, cast

import aiohttp

from .exceptions import (
    BringAuthException,
    BringEMailInvalidException,
    BringParseException,
    BringRequestException,
    BringUserUnknownException,
)
from .types import (
    BringAuthResponse,
    BringItemsResponse,
    BringListItemDetails,
    BringListItemsDetailsResponse,
    BringListResponse,
    BringNotificationsConfigType,
    BringNotificationType,
)

_LOGGER = logging.getLogger(__name__)


class Bring:
    """Unofficial Bring API interface."""

    def __init__(
        self, session: aiohttp.ClientSession, mail: str, password: str
    ) -> None:
        """Init function for Bring API."""
        self._session = session

        self.mail = mail
        self.password = password
        self.uuid = ""
        self.publicUuid = ""

        self.url = "https://api.getbring.com/rest/"

        self.headers = {
            "Authorization": "Bearer",
            "X-BRING-API-KEY": "cof4Nc6D8saplXjE3h3HXqHH8m7VU2i1Gs0g85Sp",
            "X-BRING-CLIENT": "android",
            "X-BRING-APPLICATION": "bring",
            "X-BRING-COUNTRY": "DE",
            "X-BRING-USER-UUID": "",
        }
        self.putHeaders = {
            "Authorization": "",
            "X-BRING-API-KEY": "",
            "X-BRING-CLIENT-SOURCE": "",
            "X-BRING-CLIENT": "",
            "X-BRING-COUNTRY": "",
            "X-BRING-USER-UUID": "",
            "Content-Type": "",
        }
        self.postHeaders = {
            "Authorization": "",
            "X-BRING-API-KEY": "",
            "X-BRING-CLIENT-SOURCE": "",
            "X-BRING-CLIENT": "",
            "X-BRING-COUNTRY": "",
            "X-BRING-USER-UUID": "",
            "Content-Type": "",
        }

    async def login(self) -> BringAuthResponse:
        """Try to login.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.
        BringParseException
            If the parsing of the request response fails.
        BringAuthException
            If the login fails due to missing data in the API response.
            You should check your email and password.

        """
        user_data = {"email": self.mail, "password": self.password}

        try:
            url = f"{self.url}v2/bringauth"
            async with self._session.post(url, data=user_data) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)

                if r.status == 401:
                    try:
                        errmsg = await r.json()
                    except JSONDecodeError:
                        _LOGGER.error(
                            "Exception: Cannot parse login request response:\n %s",
                            traceback.format_exc(),
                        )
                    else:
                        _LOGGER.error("Exception: Cannot login: %s", errmsg["message"])
                    raise BringAuthException(
                        "Login failed due to authorization failure, please check your email and password."
                    )
                elif r.status == 400:
                    _LOGGER.error("Exception: Cannot login: %s", await r.text())
                    raise BringAuthException(
                        "Login failed due to bad request, please check your email."
                    )
                r.raise_for_status()

                try:
                    data = cast(
                        BringAuthResponse,
                        {
                            key: val
                            for key, val in (await r.json()).items()
                            if key in BringAuthResponse.__annotations__
                        },
                    )
                except JSONDecodeError as e:
                    _LOGGER.error(
                        "Exception: Cannot login:\n %s", traceback.format_exc()
                    )
                    raise BringParseException(
                        "Cannot parse login request response."
                    ) from e
        except asyncio.TimeoutError as e:
            _LOGGER.error("Exception: Cannot login:\n %s", traceback.format_exc())
            raise BringRequestException(
                "Authentication failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error("Exception: Cannot login:\n %s", traceback.format_exc())
            raise BringRequestException(
                "Authentication failed due to request exception."
            ) from e

        if "uuid" not in data or "access_token" not in data:
            _LOGGER.error("Exception: Cannot login: Data missing in API response.")
            raise BringAuthException(
                "Login failed due to missing data in the API response, please check your email and password."
            )

        self.uuid = data["uuid"]
        self.publicUuid = data.get("publicUuid", "")
        self.headers["X-BRING-USER-UUID"] = self.uuid
        self.headers["Authorization"] = f'Bearer {data["access_token"]}'
        self.putHeaders = {
            **self.headers,
            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        }
        self.postHeaders = {
            **self.headers,
            "Content-Type": "application/json; charset=UTF-8",
        }
        return data

    async def loadLists(self) -> BringListResponse:
        """Load all shopping lists.

        Returns
        -------
        dict

        The JSON response as a dict.

        Raises
        ------
        BringRequestException
            If the request fails.
        BringParseException
            If the parsing of the request response fails.

        """
        try:
            url = f"{self.url}bringusers/{self.uuid}/lists"
            async with self._session.get(url, headers=self.headers) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()

                try:
                    data = cast(
                        BringListResponse,
                        {
                            key: val
                            for key, val in (await r.json()).items()
                            if key in BringListResponse.__annotations__
                        },
                    )
                    return data
                except JSONDecodeError as e:
                    _LOGGER.error(
                        "Exception: Cannot get lists:\n %s", traceback.format_exc()
                    )
                    raise BringParseException(
                        "Loading lists failed during parsing of request response."
                    ) from e
        except asyncio.TimeoutError as e:
            _LOGGER.error("Exception: Cannot get lists:\n %s", traceback.format_exc())
            raise BringRequestException(
                "Loading list failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error("Exception: Cannot get lists:\n %s", traceback.format_exc())
            raise BringRequestException(
                "Loading lists failed due to request exception."
            ) from e

    async def getItems(self, listUuid: str) -> BringItemsResponse:
        """Get all items from a shopping list.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()

        Returns
        -------
        dict
            The JSON response as a dict.

        Raises
        ------
        BringRequestException
            If the request fails.
        BringParseException
            If the parsing of the request response fails.

        """
        try:
            url = f"{self.url}v2/bringlists/{listUuid}"
            async with self._session.get(url, headers=self.headers) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()

                try:
                    data = cast(
                        BringItemsResponse,
                        {
                            key: val
                            for key, val in (await r.json())["items"].items()
                            if key in BringItemsResponse.__annotations__
                        },
                    )
                    return data
                except JSONDecodeError as e:
                    _LOGGER.error(
                        "Exception: Cannot get items for list %s:\n%s",
                        listUuid,
                        traceback.format_exc(),
                    )
                    raise BringParseException(
                        "Loading list items failed during parsing of request response."
                    ) from e
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot get items for list %s:\n%s",
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                "Loading list items failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot get items for list %s:\n%s",
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                "Loading list items failed due to request exception."
            ) from e

    async def getAllItemDetails(self, listUuid: str) -> BringListItemsDetailsResponse:
        """Get all details from a shopping list.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()

        Returns
        -------
        list
            The JSON response as a list. A list of item details.
            Caution: This is NOT a list of the items currently marked as 'to buy'. See getItems() for that.
            This contains the items that where customized by changing their default icon, category or uploading
            an image.

        Raises
        ------
        BringRequestException
            If the request fails.
        BringParseException
            If the parsing of the request response fails.

        """
        try:
            url = f"{self.url}bringlists/{listUuid}/details"
            async with self._session.get(url, headers=self.headers) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()

                try:
                    data = [
                        cast(
                            BringListItemDetails,
                            {
                                key: val
                                for key, val in item.items()
                                if key in BringListItemDetails.__annotations__
                            },
                        )
                        for item in (await r.json())["items"]
                    ]
                    return cast(BringListItemsDetailsResponse, data)
                except JSONDecodeError as e:
                    _LOGGER.error(
                        "Exception: Cannot get item details for list %s:\n%s",
                        listUuid,
                        traceback.format_exc(),
                    )
                    raise BringParseException(
                        "Loading list details failed during parsing of request response."
                    ) from e
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot get item details for list %s:\n%s",
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                "Loading list details failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot get item details for list %s:\n%s",
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                "Loading list details failed due to request exception."
            ) from e

    async def saveItem(
        self, listUuid: str, itemName: str, specification: str = ""
    ) -> aiohttp.ClientResponse:
        """Save an item to a shopping list.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()
        itemName : str
            The name of the item you want to save.
        specification : str, optional
            The details you want to add to the item.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.

        """
        data = {
            "purchase": itemName,
            "specification": specification,
        }
        try:
            url = f"{self.url}v2/bringlists/{listUuid}"
            async with self._session.put(url, headers=self.putHeaders, data=data) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()
                return r
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot save item %s (%s) to list %s:\n%s",
                itemName,
                specification,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Saving item {itemName} ({specification}) to list {listUuid} failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot save item %s (%s) to list %s:\n%s",
                itemName,
                specification,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Saving item {itemName} ({specification}) to list {listUuid} failed due to request exception."
            ) from e

    async def updateItem(
        self, listUuid: str, itemName: str, specification: str = ""
    ) -> aiohttp.ClientResponse:
        """Update an existing list item.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()
        itemName : str
            The name of the item you want to update.
        specification : str, optional
            The details you want to update on the item.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.

        """
        data = {"purchase": itemName, "specification": specification}
        try:
            url = f"{self.url}v2/bringlists/{listUuid}"
            async with self._session.put(url, headers=self.putHeaders, data=data) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()
                return r
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot update item %s (%s) to list %s:\n%s",
                itemName,
                specification,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Updating item {itemName} ({specification}) in list {listUuid} failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot update item %s (%s) to list %s:\n%s",
                itemName,
                specification,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Updating item {itemName} ({specification}) in list {listUuid} failed due to request exception."
            ) from e

    async def removeItem(self, listUuid: str, itemName: str) -> aiohttp.ClientResponse:
        """Remove an item from a shopping list.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()
        itemName : str
            The name of the item you want to remove.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.

        """
        data = {
            "remove": itemName,
        }
        try:
            url = f"{self.url}v2/bringlists/{listUuid}"
            async with self._session.put(url, headers=self.putHeaders, data=data) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()
                return r
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot delete item %s from list %s:\n%s",
                itemName,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Removing item {itemName} from list {listUuid} failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot delete item %s from list %s:\n%s",
                itemName,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Removing item {itemName} from list {listUuid} failed due to request exception."
            ) from e

    async def completeItem(
        self, listUuid: str, itemName: str
    ) -> aiohttp.ClientResponse:
        """Complete an item from a shopping list. This will add it to recent items.

        If it was not on the list, it will still be added to recent items.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()
        itemName : str
            The name of the item you want to complete.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.

        """
        data = {"recently": itemName}
        try:
            url = f"{self.url}v2/bringlists/{listUuid}"
            async with self._session.put(url, headers=self.putHeaders, data=data) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()
                return r
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot complete item %s in list %s:\n%s",
                itemName,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Completing item {itemName} from list {listUuid} failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot complete item %s in list %s:\n%s",
                itemName,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Completing item {itemName} from list {listUuid} failed due to request exception."
            ) from e

    async def notify(
        self,
        listUuid: str,
        notificationType: BringNotificationType,
        itemName: str,
    ) -> aiohttp.ClientResponse:
        """Send a push notification to all other members of a shared list.

        Parameters
        ----------
        listUuid : str
            A list uuid returned by loadLists()
        notificationType : BringNotificationType
            The type of notification to be sent
        itemName : str, optional
            The text that **must** be included in the URGENT_MESSAGE BringNotificationType.

        Returns
        -------
        Response
            The server response object.

        Raises
        ------
        BringRequestException
            If the request fails.

        """
        json = BringNotificationsConfigType(
            arguments=[],
            listNotificationType=notificationType.value,
            senderPublicUserUuid=self.publicUuid,
        )

        if not isinstance(notificationType, BringNotificationType):
            _LOGGER.error(
                "Exception: notificationType %s not supported.", notificationType
            )
            raise TypeError(
                f"notificationType {notificationType} not supported, must be of type BringNotificationType."
            )
        if notificationType is BringNotificationType.URGENT_MESSAGE:
            if not itemName or len(itemName) == 0:
                _LOGGER.error("Exception: Argument itemName missing.")
                raise ValueError(
                    "notificationType is URGENT_MESSAGE but argument itemName missing."
                )
            else:
                json["arguments"] = [itemName]
        try:
            url = f"{self.url}v2/bringnotifications/lists/{listUuid}"
            async with self._session.post(
                url, headers=self.postHeaders, json=json
            ) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)
                r.raise_for_status()
                return r
        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot send notification %s for list %s:\n%s",
                notificationType,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Sending notification {notificationType} for list {listUuid} failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: Cannot send notification %s for list %s:\n%s",
                notificationType,
                listUuid,
                traceback.format_exc(),
            )
            raise BringRequestException(
                f"Sending notification {notificationType} for list {listUuid} failed due to request exception."
            ) from e

    async def does_user_exist(self, mail: Optional[str] = None) -> bool:
        """Check if e-mail is valid and user exists.

        Parameters
        ----------
        mail : str
            An e-mail address.

        Returns
        -------
        bool
            True if user exists.

        Raises
        ------
        BringRequestException
            If the request fails.
        BringEMailInvalidException
            If the email is invalid.
        BringUserUnknownException
            If the user is does not exist

        """
        mail = mail or self.mail

        if not mail:
            raise ValueError("Argument mail missing.")

        params = {"email": mail}

        try:
            url = f"{self.url}bringusers"
            async with self._session.get(url, headers=self.headers, params=params) as r:
                _LOGGER.debug("Response from %s: %s", url, r.status)

                if r.status == 404:
                    _LOGGER.error("Exception: User %s does not exist.", mail)
                    raise BringUserUnknownException(f"User {mail} does not exist.")

                r.raise_for_status()

        except asyncio.TimeoutError as e:
            _LOGGER.error(
                "Exception: Cannot get verification for %s:\n%s",
                mail,
                traceback.format_exc(),
            )
            raise BringRequestException(
                "Verifying email failed due to connection timeout."
            ) from e
        except aiohttp.ClientError as e:
            _LOGGER.error(
                "Exception: E-mail %s is invalid.",
                mail,
            )
            raise BringEMailInvalidException(f"E-mail {mail} is invalid.") from e

        return True
