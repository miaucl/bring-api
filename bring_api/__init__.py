"""Bring API package."""

__VERSION__ = "0.8.1a1"

from .bring import Bring
from .exceptions import (
    BringAuthException,
    BringEMailInvalidException,
    BringParseException,
    BringRequestException,
    BringTranslationException,
    BringUserUnknownException,
)
from .types import (
    BringAuthResponse,
    BringAuthTokenRespone,
    BringItem,
    BringItemOperation,
    BringItemsResponse,
    BringListItemDetails,
    BringListItemsDetailsResponse,
    BringListResponse,
    BringNotificationsConfigType,
    BringNotificationType,
    BringSyncCurrentUserResponse,
    BringUserListSettingEntry,
    BringUserSettingsEntry,
    BringUserSettingsResponse,
)

__all__ = [
    "Bring",
    "BringAuthException",
    "BringAuthResponse",
    "BringAuthTokenRespone",
    "BringEMailInvalidException",
    "BringItem",
    "BringItemOperation",
    "BringItemsResponse",
    "BringListItemDetails",
    "BringListItemsDetailsResponse",
    "BringListResponse",
    "BringNotificationsConfigType",
    "BringNotificationType",
    "BringParseException",
    "BringRequestException",
    "BringSyncCurrentUserResponse",
    "BringTranslationException",
    "BringUserListSettingEntry",
    "BringUserSettingsEntry",
    "BringUserSettingsResponse",
    "BringUserUnknownException",
]
