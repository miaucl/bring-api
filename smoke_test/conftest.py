"""Smoke test for bring-api."""

import logging
import os

import aiohttp
from dotenv import load_dotenv
import pytest

from bring_api.bring import Bring
from bring_api.helpers import headers_deserialize, headers_serialize
from bring_api.types import BringList

load_dotenv()

_LOGGER = logging.getLogger(__name__)


def save_headers(headers: dict[str, str]) -> None:
    """Save the headers locally."""
    with open(".headers", "w", encoding="utf-8") as file:
        file.write(headers_serialize(headers))


def load_headers() -> dict[str, str]:
    """Load the headers locally."""
    # Open and read the file
    with open(".headers", encoding="utf-8") as file:
        return headers_deserialize(file.read() or "{}")


@pytest.fixture(name="headers")
async def cookies_str() -> dict[str, str]:
    """Load the headers."""

    return load_headers()


@pytest.fixture(name="session")
async def aiohttp_client_session():
    """Create  a client session."""
    async with aiohttp.ClientSession() as session:
        yield session


@pytest.fixture(name="bring")
async def bring_api_client(
    session: aiohttp.ClientSession, headers: dict[str, str]
) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    # Restore auth data from saved headers
    bring.headers = headers
    bring.uuid = headers["X-BRING-USER-UUID"]
    bring.public_uuid = headers["X-BRING-PUBLIC-USER-UUID"]
    await bring.reload_user_list_settings()
    await bring.reload_article_translations()

    return bring


@pytest.fixture(name="bring_no_auth")
async def bring_no_auth_api_client(session: aiohttp.ClientSession) -> Bring:
    """Create Bring instance."""

    bring = Bring(session, os.environ["EMAIL"], os.environ["PASSWORD"])

    return bring


@pytest.fixture(name="test_list")
async def test_list(bring: Bring) -> BringList:
    """Get the BringList instance to test with."""
    # Get information about all available shopping list and select the one to test with
    lists = (await bring.load_lists()).lists
    lst = next(lst for lst in lists if lst.name == os.environ["LIST"])
    _LOGGER.info("Selected list: %s (%s)", lst.name, lst.listUuid)
    return lst
