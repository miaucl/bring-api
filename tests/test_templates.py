"""Tests for parse_recipe."""

import pytest
from syrupy.assertion import SnapshotAssertion

from bring_api import Bring, BringTemplate

from .conftest import load_fixture


@pytest.mark.usefixtures("mocked")
async def test_parse_recipe(
    bring: Bring,
    snapshot: SnapshotAssertion,
) -> None:
    """Test parse_recipe."""
    await bring.login()
    result = await bring.parse_recipe("https://example.com")

    assert result == snapshot


@pytest.mark.usefixtures("mocked")
async def test_create_template(
    bring: Bring,
    snapshot: SnapshotAssertion,
) -> None:
    """Test create_template."""
    await bring.login()

    template = BringTemplate.from_json(load_fixture("recipe.json"))
    result = await bring.create_template(template)

    assert result == snapshot


@pytest.mark.usefixtures("mocked")
async def test_delete_recipe(
    bring: Bring,
    snapshot: SnapshotAssertion,
) -> None:
    """Test delete_recipe."""
    await bring.login()

    await bring.delete_template("98ad5860-e8d2-4e3f-8c9e-0fe3f59eec8a")
