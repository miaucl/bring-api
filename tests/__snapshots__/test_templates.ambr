# serializer version: 1
# name: test_create_template
  BringTemplate(name='Erdnusspfanne', author='Chefkoch', tagline=None, linkOutUrl='https://www.chefkoch.de/rezepte/3844601585665452/Erdnusspfanne.html', imageUrl='https://img.chefkoch-cdn.de/rezepte/3844601585665452/bilder/1296333/crop-960x720/erdnusspfanne.jpg', imageWidth=960, imageHeight=720, items=[], requestQuantity=None, baseQuantity=None, time=None, ingredients=[], likeCount=38, tags=['bring_recipe_parser', 'vegetarisch', 'Gemüse', 'chefkoch.de', 'Pilze', 'Hauptspeise', 'einfach', 'vegan', 'Schnell', 'gekocht'], enableQuantityChange=True, uuid='98ad5860-e8d2-4e3f-8c9e-0fe3f59eec8a', contentUuid='647a2a25-e158-4747-a9f4-2c5887ac840d', contentVersion=1, campaign=None, title='Erdnusspfanne', importCount=0, template_type=<TemplateType.RECIPE: 'RECIPE'>, isPromoted=None, isAd=None, contentSrcUrl='https://api.getbring.com/rest/v2/bringtemplates/content/647a2a25-e158-4747-a9f4-2c5887ac840d', attribution='', attributionSubtitle='', attributionIcon='')
# ---
# name: test_parse_recipe
  BringTemplate(name='Erdnusspfanne', author='Chefkoch', tagline='Gesamtzeit ca. 35 Minuten', linkOutUrl='https://www.chefkoch.de/rezepte/3844601585665452/Erdnusspfanne.html', imageUrl='https://img.chefkoch-cdn.de/rezepte/3844601585665452/bilder/1296333/crop-960x720/erdnusspfanne.jpg', imageWidth=960, imageHeight=720, items=[Ingredient(itemId='Tofu', stock=False, spec='1 Pck., (200 g)', altIcon=None, altSection=None), Ingredient(itemId='Paprikaschote(n)', stock=False, spec='1', altIcon='Peperoni', altSection='Früchte & Gemüse'), Ingredient(itemId='Champignons', stock=False, spec='250g', altIcon=None, altSection=None), Ingredient(itemId='Zwiebel(n)', stock=False, spec='1', altIcon='Zwiebeln', altSection='Früchte & Gemüse'), Ingredient(itemId='Knoblauchzehe(n)', stock=False, spec='1', altIcon='Knoblauch', altSection='Früchte & Gemüse'), Ingredient(itemId='Erdnusscreme', stock=False, spec='3 EL', altIcon='Nougatcreme', altSection='Snacks & Süsswaren'), Ingredient(itemId='Soyasauce', stock=False, spec='3 EL', altIcon=None, altSection=None), Ingredient(itemId='Gemüsebrühe', stock=False, spec='200ml', altIcon='Bouillon', altSection='Zutaten & Gewürze'), Ingredient(itemId='Hafermilch (Haferdrink)', stock=False, spec='300ml, oder Sojadrink', altIcon='Milch', altSection='Milch & Käse'), Ingredient(itemId='Salz und Pfeffer', stock=True, spec=None, altIcon='Salz', altSection=None), Ingredient(itemId='Currypulver', stock=False, spec=None, altIcon='Paprikapulver', altSection='Zutaten & Gewürze'), Ingredient(itemId='Chilipulver', stock=False, spec=None, altIcon='Paprikapulver', altSection='Zutaten & Gewürze'), Ingredient(itemId='Petersilie', stock=False, spec='optional', altIcon=None, altSection=None), Ingredient(itemId='Kokosöl', stock=False, spec=None, altIcon='Öl', altSection='Zutaten & Gewürze')], requestQuantity='4', baseQuantity=4, time='PT35M', ingredients=[Ingredient(itemId='Tofu', stock=False, spec='1 Pck., (200 g)', altIcon=None, altSection=None), Ingredient(itemId='Paprikaschote(n)', stock=False, spec='1', altIcon='Peperoni', altSection='Früchte & Gemüse'), Ingredient(itemId='Champignons', stock=False, spec='250g', altIcon=None, altSection=None), Ingredient(itemId='Zwiebel(n)', stock=False, spec='1', altIcon='Zwiebeln', altSection='Früchte & Gemüse'), Ingredient(itemId='Knoblauchzehe(n)', stock=False, spec='1', altIcon='Knoblauch', altSection='Früchte & Gemüse'), Ingredient(itemId='Erdnusscreme', stock=False, spec='3 EL', altIcon='Nougatcreme', altSection='Snacks & Süsswaren'), Ingredient(itemId='Soyasauce', stock=False, spec='3 EL', altIcon=None, altSection=None), Ingredient(itemId='Gemüsebrühe', stock=False, spec='200ml', altIcon='Bouillon', altSection='Zutaten & Gewürze'), Ingredient(itemId='Hafermilch (Haferdrink)', stock=False, spec='300ml, oder Sojadrink', altIcon='Milch', altSection='Milch & Käse'), Ingredient(itemId='Salz und Pfeffer', stock=True, spec=None, altIcon='Salz', altSection=None), Ingredient(itemId='Currypulver', stock=False, spec=None, altIcon='Paprikapulver', altSection='Zutaten & Gewürze'), Ingredient(itemId='Chilipulver', stock=False, spec=None, altIcon='Paprikapulver', altSection='Zutaten & Gewürze'), Ingredient(itemId='Petersilie', stock=False, spec='optional', altIcon=None, altSection=None), Ingredient(itemId='Kokosöl', stock=False, spec=None, altIcon='Öl', altSection='Zutaten & Gewürze')], likeCount=38, tags=['bring_recipe_parser', 'Gemüse', 'bring_mealDiet_vegetarian', 'bring_mealType_easy', 'chefkoch.de', 'Pilze', 'vegan', 'bring_mealType_mainCourse', 'gekocht'], enableQuantityChange=True, uuid=None, contentUuid=None, contentVersion=None, campaign=None, title=None, importCount=None, template_type=<TemplateType.TEMPLATE: 'TEMPLATE'>, isPromoted=None, isAd=None, contentSrcUrl=None, attribution=None, attributionSubtitle=None, attributionIcon=None)
# ---
